packages/iframe-execution-environment-service/src/IframeExecutionEnvironmentService.ts

```diff
@@ -0,0 +1,337 @@
+  implements ExecutionEnvironmentService
+{
+  public store: ObservableStore<{ jobs: Record<string, EnvMetadata> }>;
+
+  private _pluginRpcHooks: Map<string, PluginRpcHook>;
+
+  public _iframeWindow?: Window;
+
+  public iframeUrl: URL;
+
+  private jobs: Map<string, EnvMetadata>;
+
+  private setupPluginProvider: SetupPluginProvider;
+
+  private pluginToJobMap: Map<string, string>;
+
+  private jobToPluginMap: Map<string, string>;
+
+  constructor({ setupPluginProvider, iframeUrl }: JobControllerArgs) {
+    this.iframeUrl = iframeUrl;
+    this.setupPluginProvider = setupPluginProvider;
+    this.store = new ObservableStore({ jobs: {} });
+    this.jobs = new Map();
+    this.pluginToJobMap = new Map();
+    this.jobToPluginMap = new Map();
+    this._pluginRpcHooks = new Map();
+  }
+
+  private _setJob(jobId: string, jobWrapper: EnvMetadata): void {
+    this.jobs.set(jobId, jobWrapper);
+
+    const newJobState = {
+      ...(this.store.getState().jobs as Record<string, EnvMetadata>),
+      [jobId]: jobWrapper,
+    };
+    this.store.updateState({ jobs: newJobState });
+  }
+
+  private _deleteJob(jobId: string): void {
+    this.jobs.delete(jobId);
+
+    const newJobState = {
+      ...(this.store.getState().jobs as Record<string, EnvMetadata>),
+    };
+    delete newJobState[jobId];
+    this.store.updateState({ jobs: newJobState });
+  }
+
+  private async _command(
+    jobId: string,
+    message: JsonRpcRequest<unknown>,
+  ): Promise<unknown> {
+    if (typeof message !== 'object') {
+      throw new Error('Must send object.');
+    }
+
+    const jobWrapper = this.jobs.get(jobId);
+    if (!jobWrapper) {
+      throw new Error(`Job with id ${jobId} not found.`);
+    }
+
+    console.log('Parent: Sending Command', message);
+    const response: PendingJsonRpcResponse<unknown> =
+      await jobWrapper.rpcEngine.handle(message);
+    if (response.error) {
+      throw new Error(response.error.message);
+    }
+    return response.result;
+  }
+
+  async terminateAllPlugins() {
+    for (const jobId of this.jobs.keys()) {
+      this.terminate(jobId);
+    }
+    this._pluginRpcHooks.clear();
+  }
+
+  async terminatePlugin(pluginName: string) {
+    const jobId = this.pluginToJobMap.get(pluginName);
+    jobId && this.terminate(jobId);
+    this._removePluginHooks(pluginName);
+  }
+
+  terminate(jobId: string): void {
+    const jobWrapper = this.jobs.get(jobId);
+    if (!jobWrapper) {
+      throw new Error(`Job with id "${jobId}" not found.`);
+    }
+
+    Object.values(jobWrapper.streams).forEach((stream) => {
+      try {
+        !stream.destroyed && stream.destroy();
+        stream.removeAllListeners();
+      } catch (err) {
+        console.log('Error while destroying stream', err);
+      }
+    });
+    document.getElementById(jobWrapper.id)?.remove();
+    this._removePluginAndJobMapping(jobId);
+    this._deleteJob(jobId);
+    console.log(`job:${jobId} terminated`);
```

Comment:
In line with the above comment re: template string values, this will be difficult to parse if the job ID is ever missing. There are similar literals like this elsewhere in the file. I recommend changing them to the format on L129 above, i.e. `Job with id "${jobId}" ...`.